File: examples/cpp/test/rpg/rpgtest.cc
#include "rpg/rpg.h"
#include "rpg/inventory/inventory.h"
#include "doctest.h"

using namespace rpg;
using namespace rpg::inventory;

TEST_CASE ("CharacterWriteRead") {
  auto character = std::unique_ptr<Character>{
    Character::Builder()
      .setName("Bobby")
      .setPosition({-1, 1})
      .setSpeed(3.0f)
      .setBag(Inventory::Builder()
                .setCapacity(10)
                .setItems({
                            new Weapon{10, "A", Quality::COMMON, 5},
                            new Armor{15, "B", Quality::RARE, 15}
                          })
                .build())
      .setEquipment("MainHand", new Weapon{50, "Knife", Quality::COMMON, 10})
      .setEquipment("Head", new Armor{100, "Fedora", Quality::EPIC, 42})
      .setBuffs({32.0, 64.0})
      .build()
  };
  bufobjects::BufferBuilder bb;

  bufobjects::writeIdentifiedTo(bb, character.get());
  character->clear();
  bb.Rewind();

  character.reset(dynamic_cast<Character*>(bufobjects::readIdentifiedFrom(bb)));

    CHECK_EQ("Bobby", character->getName());
    CHECK_EQ(3.0f, character->getSpeed());
    CHECK_EQ(32.0, character->getBuffs(0));
    CHECK_EQ(64.0, character->getBuffs(1));
    CHECK_EQ(-1, character->getPosition().x);
    CHECK_EQ(1, character->getPosition().y);

    CHECK(character->hasBag());
  auto weapon = dynamic_cast<Weapon*>(character->getBag()->getItems(0));
    CHECK_EQ("A", weapon->getName());
    CHECK_EQ(5, weapon->getCost());
    CHECK_EQ(Quality::COMMON, weapon->getQuality());
    CHECK_EQ(10, weapon->getDamage());
  auto armor = dynamic_cast<Armor*>(character->getBag()->getItems(1));
    CHECK_EQ("B", armor->getName());
    CHECK_EQ(15, armor->getCost());
    CHECK_EQ(Quality::RARE, armor->getQuality());
    CHECK_EQ(15, armor->getDefense());
  weapon = dynamic_cast<Weapon*>(character->getEquipment("MainHand"));
    CHECK_EQ("Knife", weapon->getName());
    CHECK_EQ(10, weapon->getCost());
    CHECK_EQ(Quality::COMMON, weapon->getQuality());
    CHECK_EQ(50, weapon->getDamage());
  armor = dynamic_cast<Armor*>(character->getEquipment("Head"));
    CHECK_EQ("Fedora", armor->getName());
    CHECK_EQ(42, armor->getCost());
    CHECK_EQ(Quality::EPIC, armor->getQuality());
    CHECK_EQ(100, armor->getDefense());
}

File: examples/java/src/test/java/RpgTest.java
import org.junit.Assert;
import org.junit.Test;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.text.DecimalFormat;

import rpg.*;
import rpg.Character;
import rpg.inventory.*;

public class RpgTest {

  @Test
  public void testCharacterWriteRead() {
    Character character = newCharacter();
    BufferBuilder bb = new BufferBuilder();

    BufferObject.writeIdentifiedTo(bb, character);
    character.clear();
    bb.rewind();

    character = (Character) BufferObject.readIdentifiedFrom(bb);
    assertCharacter(character);
  }

  @Test
  public void testCharacterJson() throws IOException {
    Character character = newCharacter();
    ByteArrayOutputStream out = new ByteArrayOutputStream();
    character.writeJsonTo(out);

    Assert
      .assertEquals("{\"name\":\"Bobby\",\"position\":{\"x\":-1,\"y\":1},\"speed\":3,\"bag\":{\"capacity\":10,\"items\":[{\"_id\":2,\"damage\":\"10\",\"name\":\"A\",\"quality\":1,\"cost\":\"5\"},{\"_id\":3,\"defense\":\"15\",\"name\":\"B\",\"quality\":2,\"cost\":\"15\"}]},\"equipment\":{\"Head\":{\"_id\":3,\"defense\":\"100\",\"name\":\"Fedora\",\"quality\":3,\"cost\":\"42\"},\"MainHand\":{\"_id\":2,\"damage\":\"50\",\"name\":\"Knife\",\"quality\":1,\"cost\":\"10\"}},\"buffs\":[32,64,0,0,0,0,0,0]}", out
        .toString());
  }

  private Character newCharacter() {
    return Character.newBuilder()
      .setName("Bobby")
      .setPosition(new Position(-1, 1))
      .setSpeed(3.0f)
      .setBag(Inventory.newBuilder()
        .setCapacity(10)
        .setItems(new Weapon(10, "A", Quality.COMMON, 5), new Armor(15, "B", Quality.RARE, 15))
        .build())
      .setEquipment("MainHand", new Weapon(50, "Knife", Quality.COMMON, 10))
      .setEquipment("Head", new Armor(100, "Fedora", Quality.EPIC, 42))
      .setBuffs(32.0, 64.0)
      .build();
  }

  private void assertCharacter(Character character) {
    Assert.assertEquals("Bobby", character.getName());
    Assert.assertEquals(3.0f, character.getSpeed(), 0.01f);
    Assert.assertEquals(32.0, character.getBuffs(0), 0.01);
    Assert.assertEquals(64.0, character.getBuffs(1), 0.01);
    Assert.assertEquals(-1, character.getPosition().x, 0.01f);
    Assert.assertEquals(1, character.getPosition().y, 0.01f);

    Assert.assertTrue(character.hasBag());
    Weapon weapon = (Weapon) character.getBag().getItems(0);
    Assert.assertEquals("A", weapon.getName());
    Assert.assertEquals(5, weapon.getCost());
    Assert.assertEquals(Quality.COMMON, weapon.getQuality());
    Assert.assertEquals(10, weapon.getDamage());
    Armor armor = (Armor) character.getBag().getItems(1);
    Assert.assertEquals("B", armor.getName());
    Assert.assertEquals(15, armor.getCost());
    Assert.assertEquals(Quality.RARE, armor.getQuality());
    Assert.assertEquals(15, armor.getDefense());

    weapon = (Weapon) character.getEquipment("MainHand");
    Assert.assertEquals("Knife", weapon.getName());
    Assert.assertEquals(10, weapon.getCost());
    Assert.assertEquals(Quality.COMMON, weapon.getQuality());
    Assert.assertEquals(50, weapon.getDamage());
    armor = (Armor) character.getEquipment("Head");
    Assert.assertEquals("Fedora", armor.getName());
    Assert.assertEquals(42, armor.getCost());
    Assert.assertEquals(Quality.EPIC, armor.getQuality());
    Assert.assertEquals(100, armor.getDefense());
  }

}
